wasm/wasm_source/src/vp_masp.rs

```diff
@@ -70,9 +137,50 @@ fn validate_tx(
         }
 
         // Handle unshielding/transparent output
+        // The following boundary conditions must be satisfied
+        // 1. Zero transparent inupt
+        // 2. One transparent output
+        // 3. Asset type must be properly derived
+        // 4. Value from the output must be the same as the containing transfer
+        // 5. Public key must be the hash of the target
         if transfer.target != masp() {
+            // Satisfies 1.
+            if !shielded_tx.vin.is_empty() {
+                debug_log!(
+                    "Transparent input to a transaction to the masp must be 0 \
+                     but is {}",
+                    shielded_tx.vin.len()
+                );
+                return reject();
+            }
+
+            // Satisfies 2.
+            if shielded_tx.vout.len() != 1 {
+                debug_log!(
+                    "Transparent output to a transaction to the masp must be \
+                     1 but is {}",
+                    shielded_tx.vin.len()
+                );
+                return reject();
+            }
+
+            let out: &TxOut = &shielded_tx.vout[0];
+
+            let expected_asset_type: AssetType =
+                asset_type_from_epoched_address(
+                    ctx.get_block_epoch().unwrap(),
+                    &transfer.token,
+                );
+
+            // Satisfies 3. and 4.
+            if !(valid_asset_type(&expected_asset_type, &out.asset_type)
+                && valid_transfer_amount(out.value, u64::from(transfer.amount)))
+            {
+                return reject();
+            }
+
             // Timestamp is derived to allow unshields for older tokens
-            let atype =
+            let atype: &AssetType =
```

Comment:
This use of `.components().next().unwrap()` seems to be unsafe. What guarantees that `shielded_tx.value_balance` has at least one component? And when there are multiple components, what does the first extracted component mean semantically?